File: spark/eddington-spark.py
from sys import argv

from pyspark.sql import SparkSession
from pyspark.sql import functions
from pyspark.sql.window import Window


def main():
    spark = SparkSession.builder.appName("EddingtonSpark").getOrCreate()
    filename = argv[1]

    df = (spark
        .read
        .format("csv")
        .option("header", "true")
        .option("inferSchema", "true")
        .load(filename))

    window_spec = (functions.row_number()
        .over(Window.orderBy(functions.desc("sum_len"))))

    result = (df
        .groupBy("ride_date")
        .agg(functions.sum("ride_length").alias("sum_len"))
        .withColumn("row_num", window_spec)
        .where("sum_len >= row_num")
        .agg(functions.max("row_num").alias("eddington_number")))

    result.show()


if __name__ == "__main__":
    main()

File: spark/scala/src/main/scala/EddingtonSpark.scala
package bike.eddington.spark

import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.expressions.Window
import org.apache.spark.sql.functions._

object EddingtonSpark {
  def main(args: Array[String]): Unit = {
    val spark = SparkSession
      .builder
      .appName("EddingtonSpark")
      .getOrCreate()

    val filename = args(0)
    val opts = Map(
      "header" -> "true",
      "inferSchema" -> "true"
    )

    val df = spark
      .read
      .format("csv")
      .options(opts)
      .load(filename)

    val result = df
      .groupBy("ride_date")
      .agg(sum("ride_length") as "sum_len")
      .withColumn("row_num", row_number().over(Window.orderBy(desc("sum_len"))))
      .where("sum_len >= row_num")
      .agg(max("row_num") as "eddington_number")

    println(result.collect().last.getInt(0))
  }
}
